controller.h

```diff
@@ -0,0 +1,473 @@
+#pragma once
+using namespace std;
+
+// Controller
+//
+// Represents the entire system, contains the canvases which contain the features which contain the effects.
+
+#include "json.hpp"
+#include "interfaces.h"
+#include "basegraphics.h"
+#include "effectsmanager.h"
+#include <vector>
+#include "global.h"
+#include "canvas.h"
+#include "ledfeature.h"
+#include "effects/colorwaveeffect.h"
+#include "effects/starfield.h"
+#include "effects/videoeffect.h"
+#include "effects/misceffects.h"
+#include "palette.h"
+#include "effects/paletteeffect.h"
+#include "effects/fireworkseffect.h"
+
+class Controller : public IController
+{
+  private:
+
+    vector<unique_ptr<ICanvas>> _canvases;
+    uint16_t                    _port;
+
+  public:
+
+    Controller(uint16_t port = 7777) : _port(port)
+    {
+    }
+
+    vector<ICanvas*> Canvases() const override
+    {
+        vector<ICanvas*> canvases;
+        canvases.reserve(_canvases.size());
+        for (const auto& canvas : _canvases)
+            canvases.push_back(canvas.get());
+        return canvases;
+    }
+
+    uint16_t GetPort() const override
+    {
+        return _port;
+    }
+
+    bool AddFeatureToCanvas(uint16_t canvasId, unique_ptr<ILEDFeature> feature) override
+    {
+        logger->debug("Adding feature to canvas {}...", canvasId);
+
+        auto canvas = GetCanvasById(canvasId);
+        if (canvas == nullptr)
+        {
+            logger->error("Canvas {} not found in AddFeatureToCanvas.", canvasId);
+            return false;
+        }
+
+        canvas->AddFeature(std::move(feature));
+        return true;
+    }
+
+    bool RemoveFeatureFromCanvas(uint16_t canvasId, uint16_t featureId) override
+    {
+        logger->debug("Removing feature {} from canvas {}...", featureId, canvasId);
+        
+        // Find the matching canvas and remove the feature from it
+        auto canvas = GetCanvasById(canvasId);
+        if (canvas == nullptr)
+        {
+            logger->error("Canvas {} not found in RemoveFeatureFromCanvas.", canvasId);
+            return false;
+        }
+        return canvas->RemoveFeatureById(featureId);
+    }
+
+
+    // LoadSampleCanvases
+    //
+    // Until we have full load and save ability, this function will be used to load sample canvases.
+
+    void LoadSampleCanvases() override
```

Comment:
I think you've made it a lot easier for me to override your Canvas setup with mine. :)